Sprint 14

DarkRavenn opened the pull request.

DarkRavenn committed:
Убран сториборд из проекта

DarkRavenn committed:
Сверстан launchScreen и tabBar

DarkRavenn committed:
Сверстан экран Трекеры

DarkRavenn committed:
Добавлен кастомный DatePicker

DarkRavenn committed:
Добавлено отображение трекеров

DarkRavenn committed:
Добавлено добавление трекера

DarkRavenn committed:
Добавлена фильтрация трекеров через searchBar

DarkRavenn committed:
Добавлена фильтрация по дате через DatePicker

DarkRavenn committed:
доработано создание Нерегулярного события

DarkRavenn committed:
Мелкие исправления, добавлены коментарии, убраны тестовые данные

Kirchberg commented on Tracker/Application/Controllers/Base/BaseView.swift:
Было бы хорошо не использовать такие родительские классы при построении UI. 

В данном случае чем больше будет наследований от этого класса, тем больше будет соблазн на его модификацию под нужды конкретных классов, из-за этого необходимо знать детали классов, нам будет сложнее модифицировать дочерние классы и т.д.

В своих работах дядюшка Боб (Роберт Мартин) часто обсуждает проблемы, связанные с чрезмерным использованием наследования в объектно-ориентированном программировании. В книге «Чистый код» он подчеркивает, что наследование может стать источником проблем, если используется неправильно. Основная проблема заключается в том, что наследование создает жесткую связь между классами, что усложняет их модификацию и тестирование. Он рекомендует следовать принципу замещения Лисков (Liskov Substitution Principle) и использовать композицию вместо наследования, когда это возможно.

Мартин также акцентирует внимание на том, что изменения в базовом классе могут непредсказуемо повлиять на все классы-наследники. Это делает системы менее устойчивыми к изменениям и увеличивает сложность поддержки кода. Вместо этого, рекомендуется использовать интерфейсы и композицию для создания более гибкой и легко расширяемой архитектуры.

В нашем случае лучше от этого класса избавиться совсем.

Kirchberg commented on Tracker/Application/Controllers/Models/TrackerRecordModel.swift:
```diff
@@ -0,0 +1,14 @@
+
+// Структура для хранения записи о выполнении трекера на определенную дату
+struct TrackerRecord {
+    let trackerID: String
```
⚠️ В `TrackerRecord` используется тип `String` для `trackerID`, необходимо поправить на тип `UInt` или `UUID` для исправления пункта задания 8.

Kirchberg commented on Tracker/Application/Controllers/Trackers/CommonViews/Controllers/TrackerTypeSelectionViewController.swift:
```diff
@@ -0,0 +1,267 @@
+    
+    private func onUpdateSchedule(_ schedule: [Weekday]) {
+        self.schedule = schedule
+        if schedule.count == 7 {
+            self.tableOptions[1].subtitle = "Каждый день"
+        } else {
+            self.tableOptions[1].subtitle = schedule.map { item in
+                switch item {
+                case .monday: return "Пн"
+                case .tuesday: return "Вт"
+                case .wednesday: return "Ср"
+                case .thursday: return "Чт"
+                case .friday: return "Пт"
+                case .saturday: return "Сб"
+                case .sunday: return "Вс"
+                }
+            }.joined(separator: ", ")
```
Рекомендую для модели `Tracker` использовать в проперти `schedule` не напрямую `enum Weekday`, а сделать свою структуру или обертку, например с названием `Schedule`, которая будет делать следующее:
- абстрагировать операции `addWeekDay`, `removeWeekDay`, `contains`
- методы преобразования расписания в текстовое представление для отображения в `Label`
- методы сериализации расписания при сохранении в базу данных

Это также позволит разгрузить сам `enum Weekday` и оставить в нем только кейсы.

Kirchberg commented on Tracker/Application/Controllers/Trackers/CommonViews/CustomDatePicker.swift:
```diff
@@ -0,0 +1,72 @@
+
+import UIKit
+
+final class CustomDatePicker: BaseView {
```
В случае BaseView было бы хорошо делать протоколом, но не классом.

Kirchberg commented on Tracker/Application/Controllers/Trackers/CommonViews/Controllers/TrackersViewController.swift:
```diff
@@ -0,0 +1,411 @@
+
+import UIKit
+
+//var testCategories: [TrackerCategory] = [
+//    TrackerCategory(title: "Домашний уют", trackers: [
+//        Tracker(id: "0", name: "Поливать растения", color: .ypColorSelection5, emoji: "❤️", schedule: [.tuesday, .thursday, .saturday])
+//    ]),
+//    TrackerCategory(title: "Радостные мелочи", trackers: [
+//        Tracker(id: "1", name: "Кошка заслонила камеру на созвоне", color: .ypColorSelection2, emoji: "😻", schedule: [.monday, .tuesday, .wednesday, .thursday, .friday, .saturday, .sunday]),
+//        Tracker(id: "2", name: "Бабушка прислала открытку в вотсапе", color: .ypColorSelection1, emoji: "🌺", schedule: [.monday, .tuesday, .wednesday, .thursday, .friday, .saturday, .sunday]),
+//        Tracker(id: "3", name: "Свидания в апреле", color: .ypColorSelection14, emoji: "❤️", schedule: [.monday, .tuesday, .wednesday, .thursday, .friday, .saturday, .sunday])
+//    ]),
+//    TrackerCategory(title: "Самочувствие", trackers: [
+//        Tracker(id: "4", name: "Хорошее настроение", color: .ypColorSelection16, emoji: "🙂", schedule: [.monday, .tuesday, .wednesday, .thursday, .friday, .saturday, .sunday]),
+//        Tracker(id: "5", name: "Легкая тревожность", color: .ypColorSelection8, emoji: "😪", schedule: [.monday, .tuesday, .wednesday, .thursday, .friday, .saturday, .sunday])
+//    ]),
+//]
+//
+//var testCompletedTrackers: [TrackerRecord] = [
+//    TrackerRecord(trackerID: "0", date: Date()),
+//    TrackerRecord(trackerID: "1", date: Date()),
+//    TrackerRecord(trackerID: "1", date: Calendar.current.date(byAdding: .day, value: -1, to: Date())!),
+//]
```
Было бы хорошо убрать неиспользуемые строки кода не только тут, но и по всему классу

Kirchberg commented on Tracker/Application/Controllers/Trackers/CommonViews/Controllers/TrackersViewController.swift:
```diff
@@ -0,0 +1,411 @@
+        return stackView
+    }()
+    
+        private lazy var trackersCollection: UICollectionView = {
+            let collection = UICollectionView(frame: .zero, collectionViewLayout: UICollectionViewFlowLayout())
+            collection.register(TrackerCollectionCell.self, forCellWithReuseIdentifier: TrackerCollectionCell.identifier)
+            collection.register(SupplementaryView.self, forSupplementaryViewOfKind: UICollectionView.elementKindSectionHeader, withReuseIdentifier: "header")
+            collection.translatesAutoresizingMaskIntoConstraints = false
+            collection.dataSource = self
+            collection.delegate = self
+            collection.backgroundColor = .ypWhite
+            collection.showsVerticalScrollIndicator = false
+            return collection
+        }()
```
Не забывай следить за отступами, общепринятые отступы помогают читать код быстрее. 
В общем случае поможет выделить весь файл `cmd+A` и зажать `control+i`

DarkRavenn committed:
Правки по ревью:
- Изменен тип данных на UUID для Tracker.id и TrackerRecord.trackerID
- Добавлено скрытие клавиатуры при тапе за ее пределами
- Удален класс BaseView
- Убран не используемый код в TrackersViewController
- Поправлены отступы

Kirchberg commented on Tracker/Application/Controllers/Models/TrackersCollectionCell.swift:
```diff
@@ -102,7 +102,9 @@ final class TrackerCollectionCell: UICollectionViewCell {
     @objc private func onAddButtonTapped(_ button: UIButton) {
         button.isSelected.toggle()
-        self.onUpdateTrackersDoneStatus?(button.accessibilityValue ?? "", button.isSelected)
+//        self.onUpdateTrackersDoneStatus?(button.accessibilityValue ?? "", button.isSelected)
```
Было бы хорошо убрать неиспользуемый код :)

Kirchberg commented on Tracker/Helpers/Extensions/UIViewController + ext.swift:
В названии файла было бы хорошо поправить на `UIViewController + Ext`, либо есть другие альтернативы названию `UIViewController + Extexnsions`, `UIViewController + Helpers`

DarkRavenn committed:
Правки по ревью
